Derivative operations in Builder

trenouf opened the pull request:
This commit adds the derivative operations to Builder, and switches the
SPIR-V reader to use Builder for them.

I also removed an unnecessary wqm intrinsic call in the middle of the
calculation.

Change-Id: Ie71ba1d9025ffc9e0c55509c80c47b4bcbcdb401

trenouf commented:
This PR contains two commits, because it depends on an earlier still-open PR. Please review only the topmost commit.

linqun commented:
I will review it once the test is passed. please let me know if this failure is due to Jenkins agent.

trenouf commented:
This looks like a spurious failure. But I will rebase it before retriggering testing.

trenouf commented:
V2: Rebased.

amdrexu commented on builder/llpcBuilderImplArith.cpp:
```diff
@@ -526,6 +526,117 @@ Value* BuilderImplArith::CreateFSign(
+    Value* pResult = nullptr;
+    if (SupportDpp())
+    {
+        // DPP (gfx8+) version.
```
gfx8+ -> GFX8+

amdrexu commented on builder/llpcBuilderImplArith.cpp:
```diff
@@ -526,6 +526,117 @@ Value* BuilderImplArith::CreateFSign(
+    }
+    else
+    {
+        // ds_swizzle (pre-gfx8) version
```
gfx8 -> GFX8

amdrexu commented on builder/llpcBuilderImplArith.cpp:
```diff
@@ -526,6 +526,117 @@ Value* BuilderImplArith::CreateFSign(
+// =====================================================================================================================
+// Create derivative calculation on float or vector of float or half
+Value* BuilderImplArith::CreateDerivative(
+    Value*        pInValue,   // [in] Input value
```
pInValue -> pValue

amdrexu commented on builder/llpcBuilderImplArith.cpp:
```diff
@@ -526,6 +526,117 @@ Value* BuilderImplArith::CreateFSign(
+// Create derivative calculation on float or vector of float or half
+Value* BuilderImplArith::CreateDerivative(
+    Value*        pInValue,   // [in] Input value
+    bool          isY,        // False for derivative in X direction, true for Y direction
```
isY -> isDirectionY
I think it is necessary to make it more accurate

amdrexu commented on builder/llpcBuilderRecorder.cpp:
```diff
@@ -648,6 +650,17 @@ Value* BuilderRecorder::CreateSMod(
+// Create derivative calculation on float or vector of float or half
+Value* BuilderRecorder::CreateDerivative(
+    Value*        pInValue,   // [in] Input value
+    bool          isY,        // False for derivative in X direction, true for Y direction
```
isY -> isDirectionY

amdrexu commented on builder/llpcBuilderRecorder.cpp:
```diff
@@ -648,6 +650,17 @@ Value* BuilderRecorder::CreateSMod(
+// =====================================================================================================================
+// Create derivative calculation on float or vector of float or half
+Value* BuilderRecorder::CreateDerivative(
+    Value*        pInValue,   // [in] Input value
```
pInValue -> pValue

amdrexu commented on builder/llpcBuilderRecorder.h:
```diff
@@ -237,6 +238,9 @@ class BuilderRecorder final : public Builder, BuilderRecorderMetadataKinds
     Value* CreateInverseSqrt(Value* pInValue, const Twine& instName = "") override final;
 
+    // Create derivative calculation on float or vector of float or half
+    Value* CreateDerivative(Value* pInValue, bool isY, bool isFine, const Twine& instName = "") override final;
```
pInValue -> pValue

trenouf commented:
Qun also said in #166 that he wants CreateDerivative in the misc group.

amdrexu commented:
I am OK with moving it to misc group. I think Qun said this becasuse cross-lane operations are involved and it is not simple arithmetic operations any more.

trenouf commented:
V3: Rebased and addressed review comments, including moving
        CreateDerivative into misc group, and thus moving Scalarize() helper
        functions into BuilderImplBase.

This is now a single commit.

trenouf commented:
I will land this.

amdvlk-admin commented:
### Test summary for commit cf38d02661778796933439b522eb8d3d8b377409
<details open><summary>Driver commits used in build</summary><ul>

<li>METROHASH: amd-mainline 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 68b57dba33a4d922e8f1ef1b3781c2f659ffbd1c</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 331558e93794068a786bf699d3fe23bb11bac021</li>
<li>LLVM: amd-vulkan-dev f0a8af5e93d57b890687c7ced43943a0c3a167ca</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 925/925 (100.0%)</li>
<li>Failed: 0/925 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 270023/270024 (100.0%)</li>
<li><details><summary>Failed: 1/270024 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.renderpass2.dedicated_allocation.attachment_allocation.input_output.91
Stack trace: Timeout or crahsed
```
</p></details></li>


</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 366376/366376 (100.0%)</li>
<li>Failed: 0/366376 (0.0%)</li>
</ul></details>


</ul></details>

trenouf commented:
retest this please

amdvlk-admin commented:
### Test summary for commit cf38d02661778796933439b522eb8d3d8b377409
<details open><summary>Driver commits used in build</summary><ul>

<li>METROHASH: amd-mainline 2b6fee002db6cc92345b02aeee963ebaaf4c0e2f</li>
<li>PAL: dev 68b57dba33a4d922e8f1ef1b3781c2f659ffbd1c</li>
<li>SPVGEN: dev 2f31d1170e8a12a66168b23235638c4bbc43ecdc</li>
<li>XGL: dev 331558e93794068a786bf699d3fe23bb11bac021</li>
<li>LLVM: amd-vulkan-dev f0a8af5e93d57b890687c7ced43943a0c3a167ca</li>
</ul></details>

<details open><summary>ShaderDB test</summary><ul>

<li>Passed: 925/925 (100.0%)</li>
<li>Failed: 0/925 (0.0%)</li>
</ul></details>

<details open><summary>CTS tests (built with branch Vulkan-CTS-1.1.2.3-cherry-picks)</summary><ul>


<details open><summary>Ubuntu 18.04, Ellesmere</summary><ul>

<li>Passed: 366250/366250 (100.0%)</li>
<li>Failed: 0/366250 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 7.6, Vega10</summary><ul>

<li>Passed: 282963/282963 (100.0%)</li>
<li>Failed: 0/282963 (0.0%)</li>
</ul></details>


</ul></details>
